service/middleware.go

```diff
@@ -29,6 +32,35 @@ func requestIDMiddleware() mux.MiddlewareFunc {
 	}
 }
 
+func getAllowedOrigins() []string {
+	return []string{"observableusercontent.com", "localhost"}
+}
+
+func originValidator(origin string) bool {
+	valid := false
+	unescapedOrigin, err := url.QueryUnescape(origin)
+	if err != nil {
+		return valid
+	}
+	url, err := url.Parse(unescapedOrigin)
+	if err != nil {
+		return valid
+	}
+	for _, allowedDomain := range getAllowedOrigins() {
+		if strings.HasSuffix(url.Host, allowedDomain) {
+			valid = true
+		}
+	}
+	return valid
+}
+
+func corsMiddleware() mux.MiddlewareFunc {
+	return cors.New(cors.Options{
+		AllowOriginFunc:  originValidator,
+		AllowCredentials: true,
+	}).Handler
+}
```

Comment:
I think you should be able to get away with this:
```suggestion
func corsMiddleware() mux.MiddlewareFunc {
	return cors.New(cors.Options{
		AllowCredentials: true,
		AllowedMethods: []string{
			http.MethodHead,
			http.MethodGet,
			http.MethodPost,
			http.MethodPut,
			http.MethodPatch,
			http.MethodDelete,
		},
		AllowedOrigins:  allowedOrigins,
	}).Handler
}
```

Reply:
Much better!

Reply:
I think we still need the validator func because each observable notebook's origin has a dynamically generated sub-domain. I'll fix the suffix issue you caught though @nwt 👌

Reply:
@mason-fish then use wildcards for the observablehq domain name.

Reply:
e.g. `var allowedOrigins = []string{"*.observableusercontent.com", "localhost"}

Reply:
Oh neato, hadn't noticed the rc/cors supported that 👍

Reply:
Actually now that I'm thinking about it, if auth isn't enabled we should just allow cors on all domains. Some
```suggestion
func corsMiddleware(authEnabled bool) mux.MiddlewareFunc {
	opts := cors.Options{
		AllowCredentials: authEnabled,
		AllowedMethods: []string{
			http.MethodHead,
			http.MethodGet,
			http.MethodPost,
			http.MethodPut,
			http.MethodPatch,
			http.MethodDelete,
		}
	}
	if opts.AllowCredentials {
		opts.AllowedOrigins = allowedOrigins
	}
	return cors.New(opts).Handler
}
```

Reply:
I thought about it, disregard my last suggestion. When we eventually add a cli option people can configure this to their hearts content.